From c84fb752042f5ef6271f475348300af1c59acf51 Mon Sep 17 00:00:00 2001
From: yliu <yliu@apache.org>
Date: Tue, 29 Sep 2015 22:06:20 +0800
Subject: [PATCH 1654/2748] HDFS-9141. Thread leak in Datanode#refreshVolumes.
 (Uma Maheswara Rao G via yliu)

(cherry picked from commit 28805cb8d7b117892df22792ffa9ad375297b023)

Change-Id: Iaf28ed7ea1f456da92e10603b90a71d54739ad9b
---
 .../hadoop/hdfs/server/datanode/DataNode.java      |    9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java
index 5d1fa7f..30ad7df 100644
--- a/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java
+++ b/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataNode.java
@@ -588,7 +588,7 @@ ChangedVolumes parseChangedVolumes(String newVolumes) throws IOException {
   private synchronized void refreshVolumes(String newVolumes) throws IOException {
     Configuration conf = getConf();
     conf.set(DFS_DATANODE_DATA_DIR_KEY, newVolumes);
-
+    ExecutorService service = null;
     int numOldDataDirs = dataDirs.size();
     ChangedVolumes changedVolumes = parseChangedVolumes(newVolumes);
     StringBuilder errorMessageBuilder = new StringBuilder();
@@ -611,8 +611,8 @@ private synchronized void refreshVolumes(String newVolumes) throws IOException {
         for (BPOfferService bpos : blockPoolManager.getAllNamenodeThreads()) {
           nsInfos.add(bpos.getNamespaceInfo());
         }
-        ExecutorService service = Executors.newFixedThreadPool(
-            changedVolumes.newLocations.size());
+        service = Executors
+            .newFixedThreadPool(changedVolumes.newLocations.size());
         List<Future<IOException>> exceptions = Lists.newArrayList();
         for (final StorageLocation location : changedVolumes.newLocations) {
           exceptions.add(service.submit(new Callable<IOException>() {
@@ -661,6 +661,9 @@ public IOException call() {
         throw new IOException(errorMessageBuilder.toString());
       }
     } finally {
+      if (service != null) {
+        service.shutdown();
+      }
       conf.set(DFS_DATANODE_DATA_DIR_KEY,
           Joiner.on(",").join(effectiveVolumes));
       dataDirs = getStorageLocations(conf);
-- 
1.7.9.5

